File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        if nums is None or len(nums) == 0:
            return list()
        
        rp = 1
        arr = [1 for i in range(len(nums))]


        for i in range(1, len(nums)):
            arr[i] = rp * nums[i - 1]
            rp = arr[i]
        
        # print(arrLHS)
        rp = 1
        for i in range(len(nums) - 2, -1, -1):
            arr[i] = arr[i] * rp * nums[i + 1]
            rp = rp * nums[i + 1]
        return arr

#Time Complexity: O(m+n)
#Space complexity: O(m + m)

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        if(nums == null || nums.length == 0){
            return new int[0];
        }
        
        int[] result = new int[nums.length];
        
        for(int i = 0; i < nums.length; i++){
            result[i] = 1;
        }

        int rp = 1;
        
        for(int i = 1; i< nums.length; i++){
            rp = rp * nums[i - 1];            
            result[i] = rp;
        }
        
        rp = 1;
        
        for(int i = nums.length - 2; i >= 0; i--){
            rp = rp * nums[i + 1];
            result[i] = result[i] * rp;
        }
        
        return result;
            
           
    }
}